src/tools/snapshot/nodes/MessageSerializationNode.java

```diff
@@ -0,0 +1,454 @@
+
+    SPromise.getPromiseClass().serialize(prom, sb);
+    SResolver.getResolverClass().serialize(resolver, sb);
+
+    sb.putLongAt(base + COMMONALITY_BYTES, sb.getObjectPointer(resolver));
+    sb.putLongAt(base + COMMONALITY_BYTES + Long.BYTES, sb.getObjectPointer(prom));
+    sb.putIntAt(base + COMMONALITY_BYTES + Long.BYTES + Long.BYTES, fsender.getActorId());
+    base += COMMONALITY_BYTES + Long.BYTES + Long.BYTES + Integer.BYTES;
+
+    doArguments(args, base, sb);
+
+    return start;
+  }
+
+  @Specialization(guards = "dm.isDelivered()")
+  protected long doPromiseMessageNoResolver(final PromiseSendMessage dm,
+      final SnapshotBuffer sb) {
+
+    SPromise prom = dm.getPromise();
+    TracingActor fsender = (TracingActor) dm.getFinalSender();
+    Object[] args = dm.getArgs();
+
+    int payload = COMMONALITY_BYTES + Long.BYTES + Integer.BYTES + 1
+        + (args.length * Long.BYTES);
+    int base = sb.addMessage(payload);
+    long start = base - SnapshotBuffer.CLASS_ID_SIZE;
+
+    doCommonalities(MessageType.PromiseMessageNR, dm.getSelector(),
+        (TracingActor) dm.getSender(), base, sb);
+
+    SPromise.getPromiseClass().serialize(prom, sb);
+
+    sb.putLongAt(base + COMMONALITY_BYTES, sb.getObjectPointer(prom));
+    sb.putIntAt(base + COMMONALITY_BYTES + Long.BYTES, fsender.getActorId());
+    base += COMMONALITY_BYTES + Long.BYTES + Integer.BYTES;
+
+    doArguments(args, base, sb);
+
+    return start;
+  }
+
+  @Specialization(guards = {"!dm.isDelivered()", "dm.getResolver() != null"})
+  protected long doUndeliveredPromiseMessage(final PromiseSendMessage dm,
+      final SnapshotBuffer sb) {
+
+    SResolver resolver = dm.getResolver();
+    Object[] args = dm.getArgs();
+
+    int payload = COMMONALITY_BYTES + Long.BYTES + 1 + (args.length * Long.BYTES);
+    int base = sb.addMessage(payload);
+    long start = base - SnapshotBuffer.CLASS_ID_SIZE;
+
+    SResolver.getResolverClass().serialize(resolver, sb);
+
+    doCommonalities(MessageType.UndeliveredPromiseMessage, dm.getSelector(),
+        (TracingActor) dm.getSender(), base, sb);
+    sb.putLongAt(base + COMMONALITY_BYTES, sb.getObjectPointer(resolver));
+    base += COMMONALITY_BYTES + Long.BYTES;
+
+    doArguments(args, base, sb);
+
+    return start;
+  }
+
+  @Specialization(guards = "!dm.isDelivered()")
+  protected long doUndeliveredPromiseMessageNoResolver(final PromiseSendMessage dm,
+      final SnapshotBuffer sb) {
+    Object[] args = dm.getArgs();
+
+    int payload = COMMONALITY_BYTES + 1 + (args.length * Long.BYTES);
+    int base = sb.addMessage(payload);
+    long start = base - SnapshotBuffer.CLASS_ID_SIZE;
+
+    doCommonalities(MessageType.UndeliveredPromiseMessageNR, dm.getSelector(),
+        (TracingActor) dm.getSender(), base, sb);
+    base += COMMONALITY_BYTES;
+
+    doArguments(args, base, sb);
+
+    return start;
+  }
+
+  @Override
+  public EventualMessage deserialize(final DeserializationBuffer bb) {
```

Comment:
I am really not a fan of such large methods.
Can't you break the cases into nicely named methods?